private/pkg/manifest/module.go

```diff
@@ -245,3 +228,73 @@ func NewMemoryBlobFromReaderWithDigester(content io.Reader, digester Digester) (
 		content: contentInMemory.Bytes(),
 	}, nil
 }
+
+// BlobEqual returns true if blob a is the same as blob b. The digest is
+// checked for equality and the content bytes compared.
+//
+// An error is returned if an unexpected I/O error occurred when opening,
+// reading, or closing either blob.
+func BlobEqual(ctx context.Context, a, b Blob) (_ bool, retErr error) {
+	const blockSize = 4098
+	if !a.Digest().Equal(*b.Digest()) {
+		// digests don't match
+		return false, nil
+	}
+	aFile, err := a.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, aFile.Close()) }()
+	bFile, err := b.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, bFile.Close()) }()
+	// Read blockSize from a, then from b, and compare.
+	aBlock := make([]byte, blockSize)
+	bBlock := make([]byte, blockSize)
+	for {
+		aN, aErr := aFile.Read(aBlock)
+		bN, bErr := io.ReadAtLeast(bFile, bBlock[:aN], aN) // exactly aN bytes
+		// We're running unexpected error processing (not EOF) before comparing
+		// bytes because it doesn't matter if the returned bytes match if an
+		// error occurred before an expected EOF.
+		if bErr == io.ErrUnexpectedEOF {
+			// b is shorter; we can error early
+			return false, nil
+		}
+		if aErr != nil && aErr != io.EOF {
+			// unexpected read error
+			return false, aErr
+		}
+		if bErr != nil && bErr != io.EOF {
+			// unexpected read error
+			return false, bErr
+		}
+		if !bytes.Equal(aBlock[:aN], bBlock[:bN]) {
+			// Read content doesn't match. Don't forget to forward any errors
+			// returned.
+			return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
+		}
+		if aErr == io.EOF || bErr == io.EOF {
+			// EOF
+			break
+		}
+	}
+	aN, aErr := aFile.Read(aBlock[:1])
+	bN, bErr := bFile.Read(bBlock[:1])
+	if aN == 0 && bN == 0 && aErr == io.EOF && bErr == io.EOF {
+		// a and b are at EOF with no more data for us
+		return true, nil
+	}
+	// either a or b are longer
+	return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
```

Comment:
Small nit: keep happy path to left.
```suggestion
	if aN != 0 || bN != 0 || aErr != io.EOF || bErr != io.EOF {
		// either a or b are longer
		return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
	}
	// a and b are at EOF with no more data for us
	return true, nil
```

Reply:
Normally I would put the positive case at the end of a func, but this long expression is easier to read in the positive form ("if all these things are true, we're good to go").